### ISSUE TITLE
DATAURI not working in locations such as overlays= and elsewhere...

### ISSUE THREAD

**Issue by gwledig** [wesnoth/wesnoth#4191]
DATAURI ([https://wiki.wesnoth.org/DataURI]) is indicated to be useful for including artwork when the user hasn't downloaded a mod, e.g. for multiplayer sessions when players join, the DATAURI works currently in [option] image= but not elsewhere, e.g. in [unit] overlays= there is some discussion on the Wesnoth forums that this may be related to commas conflicting use where commas used in the DATAURI are being misinterpreted as overlay images for instance, this issue may be present in other image=paths such as [unit_type] and image_mod. It is interesting that it works in [option] image= presumably because this doesn't process commas like overlays. This does look like a but at the moment, especially as it works for [option] or the DATAURI page may need clarifying, if it could be used this would be very helpful for multiplayer mods, thanks. Gwledig.

**Comment by Pentarctagon** [comment]
@AI0867 btw

**Comment by gwledig** [comment]
hmm do I add @AI0867  to the title of the post?

**Comment by gfgtdf** [comment]
This is somehow a design flaw but not that big of an issue I think since 1.14 will autoamticially download needed add-ons if it's configured correctly when entering a multiplayer game.

**Comment by Pentarctagon** [comment]
The [related forum thread](https://forums.wesnoth.org/viewtopic.php?f=4&t=50263) as well, for future reference.  [This post](https://forums.wesnoth.org/viewtopic.php?f=4&t=50263#p645493) in particular.

**Comment by gwledig** [comment]
Well one idea I am testing is to just declare the addon as its own dependency in the PBL (e.g. dependencies="addonfoldername" so joining users then are forced to DL the mod right in MP lobby although I'm asking some views on this in the forum in case there are any views/problems about this...

**Comment by ProditorMagnus** [comment]
That is done with require_scenario=yes instead.

**Comment by gwledig** [comment]
ah so dependency stuff is not used for auto downloading.... thx

**Comment by gwledig** [comment]
That seems to work great, obvioulsy each senario has to have this declared another useful one is the version number requirement  e.g.
`require_scenario="yes"
addon_min_version="3.9.8"`
and it has to be placed within multiplayer...

this is great, but one suggestion, instead of booting the joiner to the MP game back into the lobby/game list, it would be good to automatically out the in the game itself...

**Comment by soliton-** [comment]
AFAICT this is not an issue with DataURI but with certain WML keys supporting some syntax and no way to escape characters. If there is no way to use a comma without some syntax meaning in a WML key then naturally you cannot usefully put a data URI in it.

**Comment by AI0867** [comment]
Well, [unit]overlays already uses parenthetical_split to not split on commas in IPFs, so it looks like a flag to also ignore the commas in data URIs should be sufficient.

**Comment by gwledig** [comment]
Ah is there some way to escape the commas in a single datauri path now? I've tried single quotes and a few other standard ways but no luck...

**Comment by AI0867** [comment]
No, but I'm working on a new function to handle the comma-separating that is dataURI-aware.

**Comment by gwledig** [comment]
perhaps it would be possible to allow defining DATAURI images so they can be used within image paths like a macro in case you want to re-use several times in an addon, to avoid having to include the long DATAURI string many times, potentially increasing size of addon unnecessarily e.g. 
```
#define myimage1
[datauri string]
#enddef
```
image="{myimage1}"

not tried this but the above might even work for [option] image= where DATAURI works in that context now.

**Comment by ProditorMagnus** [comment]
Macro does not reduce size of addon - it can decrease size of addon on your hard disk, but what is sent across network to people you play with is preprocessed WML. Though that WML is also compressed, which might reduce the harm of macros.

**Comment by CelticMinstrel** [comment]
If it's using parenthetical split, then surrounding the entire URI in parentheses should suffice to escape the comma - `image="(data:image/png;base64,...)"`. Does that work?

**Comment by gwledig** [comment]
I think I tried that but will have another look...

**Comment by AI0867** [comment]
Unfortunately the HDD of my development machine gave out, and I don't have a proper development environment set up right now, so it may take a while for me to get back to this.

**Comment by gwledig** [comment]
No problem thx for letting me know
